Plugin.php

```diff
@@ -467,79 +903,380 @@ public static function getFunctionParams(
 			return [];
 		}
 
+		if ( ! $is_invoke && $hook['deprecated'] === true && $code_location ) {
+			IssueBuffer::accepts(
+				new DeprecatedHook(
+					'Hook "' . $hook_name . '" is deprecated',
+					$code_location
+				),
+				$statements_source->getSuppressedIssues()
+			);
+		}
+
+		// don't modify the param types for utility types, since the stubbed ones are fine
+		if ( $is_utility ) {
+			return null;
+		}
+
 		// Check how many args the filter is registered with.
-		/** @var int */
-		$num_args = $call_args[3]->value->value ?? 1;
-		// Limit the required type params on the hook to match the registered number.
-		$hook_types = array_slice( $hook['types'], 0, $num_args );
+		if ( $is_invoke ) {
+			// need to deduct 1, since the first argument (string) is the hardcoded hook name, which is added manually later on, since it's not in the PHPDoc
+			$num_args = count( $call_args ) - 1;
+		} else {
+			$num_args = isset( $call_args[ 3 ]->value->value ) ? max( 0, (int) $call_args[ 3 ]->value->value ) : 1;
+		}
 
-		$hook_params = array_map( function ( Union $type ) : FunctionLikeParameter {
-			return new FunctionLikeParameter( 'param', false, $type, null, null, false );
+		// if the PHPDoc is missing from the do_action/apply_filters, the types can be empty - we assign "mixed" when loading stubs in that case though to avoid this
+		// this means these hooks really do not accept any args and the arg
+		// this will cause "InvalidArgument" error for add_action/add_filter and TooManyArguments error for do_action/apply_filters
+		// except where it actually has 0 args
+		if ( empty( $hook['types'] ) && $num_args !== 0 ) {
+			if ( $is_invoke ) {
+				// impossible, as it would have been populated with "mixed" already
+				// kept for completeness sake
+				$hook_types = array_fill( 0, $num_args, Type::getMixed() );
+			} else {
+				// add_action default has 1 arg, but this hook does not accept any args
+				if ( $is_action ) {
+					if ( $code_location ) {
+						IssueBuffer::accepts(
+							new HookInvalidArgs(
+								'Hook "' . $hook_name . '" does not accept any args, but the default number of args of add_action is 1. Please pass 0 as 4th argument',
```

Comment:
This seems like a very bizarre error to me. Is the expectation that every have to do `add_action( 'init', 'my_plugin_init', 10, 0 )`?

Reply:
Ah I see-- do we inspect the hook though to see if it _will_ actually pass an arg, rather than just relying on the default? E.g. I think "init" doesn't pass any args